### PR TITLE
Add and use new singleton class

### PR DESCRIPTION
**Pull request by Flamefire**
Another take on fixing #50 in a different way than #51:

Use a very basic implementation and explicit inst() calls for access

This makes it clear where singletons are used and avoids global instances, unnamed namespaces etc which can be a problem. A singleton is already a global instance.

### CODE REVIEW

File: include/turtle/detail/singleton.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// Add a private ctor to the type to prevent misuse
+#define MOCK_SINGLETON_CONS( type )           \
+private:                                      \
+friend class ::mock::detail::singleton<type>; \
+type() {}
+
+} // detail
+} // mock
+
+#endif // MOCK_SINGLETON_HPP
```
Comment (mat007) [inline include/turtle/detail/singleton.hpp]: There is actually a check by `boost-inspect` on this:
```
1 files that don't end with a newline
```

File: include/turtle/detail/mutex.hpp [hunk 2]
```diff
@@ -110,25 +108,25 @@ namespace detail
     {
         static Result abort()
         {
-            scoped_lock _( error_mutex );
+            scoped_lock _( error_mutex::inst() );
```
Comment (mat007) [inline include/turtle/detail/mutex.hpp]:
```
In file included from test_sequence.cpp:10:
In file included from ../include/turtle/sequence.hpp:13:
In file included from ../include/turtle/detail/sequence_impl.hpp:13:
../include/turtle/detail/mutex.hpp:111:28: error: use of undeclared identifier 'error_mutex'; did you mean 'error_mutex_t'?
            scoped_lock _( error_mutex::inst() );
                           ^~~~~~~~~~~
```
and similar errors reported clang https://travis-ci.org/mat007/turtle/jobs/455635617

File: test/mock_error.hpp [hunk 3]
```diff
@@ -10,12 +10,12 @@
...
 
 #define MOCK_ERROR_POLICY mock_error
+#include <turtle/detail/singleton.hpp>
 #include <boost/lexical_cast.hpp>
 #include <boost/test/unit_test.hpp>
-#include <boost/test/utils/trivial_singleton.hpp>
 #include <stdexcept>
 
-struct mock_error_data_t : boost::unit_test::singleton< mock_error_data_t >
+struct mock_error_data_t : turtle::detail::singleton< mock_error_data_t >
```
Comment (mat007) [inline test/mock_error.hpp]: g++ doesn't like this it seems https://travis-ci.org/mat007/turtle/jobs/455635618
```
In file included from test_sequence.cpp:9:0:
mock_error.hpp:18:28: error: ‘turtle’ has not been declared
 struct mock_error_data_t : turtle::detail::singleton< mock_error_data_t >
```

File: include/turtle/detail/mutex.hpp (hunk 2 above)
Comment (Flamefire) [reply include/turtle/detail/mutex.hpp]: Ouch, you are right. Should have been `error_mutex_t::inst()`

File: include/turtle/detail/singleton.hpp (hunk 1 above)
Comment (Flamefire) [reply include/turtle/detail/singleton.hpp]: Will add

File: test/mock_error.hpp (hunk 3 above)
Comment (Flamefire) [reply test/mock_error.hpp]: Ah yes, why is the namespace of `turtle/detail/foo.hpp` `mock::detail::foo`? Shouldn't the toplevel namespace be `turtle`?

Changed it now to `mock`

**Comment by Flamefire** [comment]
Only 1 failure with clang for boost 1.58: https://travis-ci.org/mat007/turtle/jobs/456014965

Any idea?

**Comment by mat007** [comment]
From a quick glance this looks like a race, I have re-started the job to see if it's spurious or happens every time.

**Comment by Flamefire** [comment]
Gone, so it is a race and spurious. Seems like it is not that thread safe as it should? Maybe run the race-checker tool over it.

Anyway: This seems to work now, anything left?

File: include/turtle/detail/singleton.hpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+#ifndef MOCK_SINGLETON_HPP
+#define MOCK_SINGLETON_HPP
+
+#include <boost/config.hpp>
+
+namespace mock {
+namespace detail {
+
+template<typename Derived>
```
Comment (mat007) [inline include/turtle/detail/singleton.hpp]: Missing spaces around `typerame Derived`

File: include/turtle/detail/group.hpp [hunk 5]
```diff
@@ -22,6 +22,13 @@ namespace detail
     class group
     {
     public:
+        ~group()
+        {
+            for( verifiables_cit it = verifiables_.begin();
+                    it != verifiables_.end(); ++it )
+                (*it)->context_destroyed();
```
Comment (mat007) [inline include/turtle/detail/group.hpp]: Is this part of the new singleton implementation or another fix?

File: include/turtle/detail/verifiable.hpp [hunk 6]
```diff
@@ -25,6 +25,8 @@ namespace detail
         virtual bool verify() const = 0;
 
         virtual void reset() = 0;
+
+        virtual void context_destroyed() = 0;
```
Comment (mat007) [inline include/turtle/detail/verifiable.hpp]: Not overly fond of the name of this function, can't we find something more in line with the rest of the interface here?

File: include/turtle/detail/function_impl_template.hpp [hunk 7]
```diff
@@ -255,6 +255,11 @@ namespace detail
             context_ = &c;
         }
 
+        virtual void context_destroyed()
+        {
+            context_ = NULL;
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply include/turtle/detail/function_impl_template.hpp]: Oh really? Using NULL makes transition to C++11 easier but if you want I change it

File: include/turtle/detail/group.hpp (hunk 5 above)
Comment (Flamefire) [reply include/turtle/detail/group.hpp]: Both. The new singleton shows or introduces a different initialization and hence finalization order. E. G. A global mocked function gets constructed before the `global` singleton so destruction uses the destroyed singleton

File: include/turtle/detail/verifiable.hpp (hunk 6 above)
Comment (Flamefire) [reply include/turtle/detail/verifiable.hpp]: Ideas? I'm not fond of it either. Better would be to initialize the context member in the ctor which forces correct order. But that would be more changes.

**Comment by Flamefire** [comment]
Unfortunately the singleton class was removed. Only some macros were kept. As the class is very simple I just implemented it here. The changes come from not using a global instance anymore (`foo_t& foo=foo_t::inst() `) because I had bad experiences with that (boost serialization had a bug with that which was hard to find)

File: include/turtle/detail/function_impl_template.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mat007) [reply include/turtle/detail/function_impl_template.hpp]: Yes, sorry, I'd rather keep the same coding style across the whole project.

**Comment by Flamefire** [comment]
Implemented requested changes: https://github.com/mat007/turtle/pull/55/commits/9e5cbee3765712690f1db8d0d398477e04887b42

**Comment by Flamefire** [comment]
@mat007 Anything missing here?

**Comment by mat007** [comment]
I wanted to look at `context_destroyed` a bit and see if we could not make that simpler/safer, but could find the time for now.
I have the feeling it's a bit odd that the call removing the context does not originate from the same place as the one adding it.

**Comment by Flamefire** [comment]
You are right. The flow is as follows:

 - a `verifiable` adds itself to a context: https://github.com/mat007/turtle/blob/e92f52821da312c5e30b447c8046893e576aefca/include/turtle/detail/function_impl_template.hpp#L254
- it stores a pointer to that context: https://github.com/mat007/turtle/blob/e92f52821da312c5e30b447c8046893e576aefca/include/turtle/detail/function_impl_template.hpp#L255
- The context stores this verifiable in a group: https://github.com/mat007/turtle/blob/e92f52821da312c5e30b447c8046893e576aefca/include/turtle/detail/root.hpp#L46 and https://github.com/mat007/turtle/blob/e92f52821da312c5e30b447c8046893e576aefca/include/turtle/detail/object_impl.hpp#L49
- On destruction of the `verifiable` it may access the context: https://github.com/mat007/turtle/blob/e92f52821da312c5e30b447c8046893e576aefca/include/turtle/detail/function_impl_template.hpp#L60
- But the context may be destroyed earlier leading to the use-after-free

To put the call in `group` rather than the context impl was done because `group` does not provide an iterable over its elements, but this might not be enough, see below

So maybe better solutions:

- When context is stored (not done in all cases) it registers itself for notification on context destruction
- Provide a shared_ptr instead of a reference and store a weak ptr -> Problem for singletons

I just found a `verifiable` might be added via 2 functions: To `group` or `group` and `children`: https://github.com/mat007/turtle/blob/e92f52821da312c5e30b447c8046893e576aefca/include/turtle/detail/function_impl_template.hpp#L252

I wonder if that code is correct: This function looks like it can be called multiple times. But only the ptr to the last context is stored, so if you call it twice with different contexts it will only be removed from the last context.

**Comment by Flamefire** [comment]
I'll leave solving this problem for later. See #62 for a fix to the singleton issue.

**Pull request closed without merging**